File: backend/src/protocol/handlers/client_ready.ts
import { EnforcedStateName } from "@/enums/EnforcedStateName.ts";
import { ErrorCode } from "@/enums/ErrorCode.ts";
import { SceneName } from "@/enums/SceneName.ts";
import { defineHandler } from "@/lib/defineHandler.ts";

export const clientReady = defineHandler(
    "client_ready",
    async (_message, conn, ws) => {
        if (!conn.isInPlay()) {
            return;
        }

        if (conn.clientReady) {
            conn.send(ws, {
                id: "error",
                data: {
                    message: "Client is already ready.",
                    code: ErrorCode.CLIENT_ALREADY_READY,
                    fatal: false,
                },
            });
            return;
        }

        conn.clientReady = true;

        if (!conn.user.onboarded) {
            conn.setEnforcedState(ws, EnforcedStateName.CAN_HOME, false);

            conn.send(ws, {
                id: "force_scene",
                data: {
                    scene: SceneName.INTRO,
                },
            });
        }
    },
);

File: backend-next/internal/app/realtime/handlers/client_ready.go
package handlers

import (
	"context"

	"github.com/floffah/echoform/backend-next/internal/app/realtime"
	"github.com/floffah/echoform/backend-next/internal/app/realtime/protocol"
)

func (h *Handler) HandleClientReady(ctx context.Context, session *realtime.Session, _ protocol.ClientReady) error {
	if !session.IsInPlay() {
		return nil
	}

	if session.ClientReady {
		return h.sendError(ctx, session, protocol.ServerErrorCodeClientAlreadyReady, "Client is already ready.", false)
	}

	session.ClientReady = true

	if session.User.Onboarded != nil && *session.User.Onboarded {
		return nil
	}

	if err := session.SetEnforcedState(ctx, protocol.ServerSetEnforcedStateNameCanHome, false); err != nil {
		return err
	}

	return session.Send(ctx, protocol.ServerForceScenePacket{
		Data: protocol.ServerForceScene{
			Scene: protocol.ServerForceSceneSceneIntro,
		},
	})
}
